File: 3d-manager-backend/src/main/java/com/manager3d/entity/DataSource.java
package com.manager3d.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;

@Data
@Entity
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "data_source")
public class DataSource {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 200)
    private String name;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 50)
    private Type type;

    @Column(nullable = false, length = 200)
    private String host;

    @Column(nullable = false)
    private Integer port;

    @Column(name = "database_name", length = 200)
    private String databaseName;

    @Column(nullable = false, length = 100)
    private String username;

    @Column(nullable = false, length = 500)
    private String password;

    @Column(name = "pool_config", columnDefinition = "TEXT")
    private String poolConfig;

    @Column(name = "ssl_enabled")
    @Builder.Default
    private Boolean sslEnabled = false;

    @Enumerated(EnumType.STRING)
    @Column(length = 20)
    @Builder.Default
    private Status status = Status.ACTIVE;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "owner_id", nullable = false)
    private User owner;

    @Enumerated(EnumType.STRING)
    @Column(length = 20)
    @Builder.Default
    private Visibility visibility = Visibility.PRIVATE;

    @Column(name = "last_test_time")
    private LocalDateTime lastTestTime;

    @Column(name = "last_test_result")
    private Boolean lastTestResult;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public enum Type {
        MYSQL, POSTGRESQL, ORACLE, SQLSERVER, DM
    }

    public enum Status {
        ACTIVE, INACTIVE, ERROR
    }

    public enum Visibility {
        PRIVATE, DEPARTMENT_SHARED, PUBLIC
    }
}

File: 3d-manager-frontend/src/types/datasource.ts
export type DataSourceType = 'MYSQL' | 'POSTGRESQL' | 'ORACLE' | 'SQLSERVER' | 'DM'
export type DataSourceStatus = 'ACTIVE' | 'ERROR' | 'DISABLED'
export type DataSourceVisibility = 'PRIVATE' | 'DEPARTMENT_SHARED' | 'PUBLIC'

export interface DataSource {
  id: number
  name: string
  type: DataSourceType
  host: string
  port: number
  databaseName: string
  username: string
  visibility: DataSourceVisibility
  status: DataSourceStatus
  sslEnabled: boolean
  poolConfig: string | null
  lastTestTime: string | null
  lastTestResult: boolean | null
  ownerId: number
  ownerName: string
  createdAt: string
  updatedAt: string
}

export interface DataSourceCreateRequest {
  name: string
  type: DataSourceType
  host: string
  port: number
  databaseName: string
  username: string
  password: string
  poolConfig?: string
  sslEnabled?: boolean
  visibility?: DataSourceVisibility
}

export interface DataSourceTestRequest {
  type: DataSourceType
  host: string
  port: number
  databaseName: string
  username: string
  password: string
  sslEnabled?: boolean
}

export interface ConnectionTestResult {
  success: boolean
  message: string
  databaseProductName?: string
  databaseProductVersion?: string
}

export interface TableInfo {
  name: string
  type: string
  remarks: string | null
}

export interface ColumnInfo {
  name: string
  type: string
  size: number
  nullable: boolean
  remarks: string | null
  defaultValue: string | null
}

export interface DataPreviewResult {
  columns: string[]
  rows: Record<string, unknown>[]
  total: number
}
